airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/operation/BigQueryStorageOperation.kt

```diff
@@ -29,39 +30,83 @@ abstract class BigQueryStorageOperation<Data>(
     protected val datasetLocation: String
 ) : StorageOperation<Data> {
     private val existingSchemas = ConcurrentHashMap.newKeySet<String>()
-    override fun prepareStage(streamId: StreamId, destinationSyncMode: DestinationSyncMode) {
+    override fun prepareStage(streamId: StreamId, suffix: String, replace: Boolean) {
         // Prepare staging table. For overwrite, it does drop-create so we can skip explicit create.
-        if (destinationSyncMode == DestinationSyncMode.OVERWRITE) {
-            truncateStagingTable(streamId)
+        if (replace) {
+            truncateStagingTable(streamId, suffix)
         } else {
-            createStagingTable(streamId)
+            createStagingTable(streamId, suffix)
         }
     }
 
-    private fun createStagingTable(streamId: StreamId) {
-        val tableId = TableId.of(streamId.rawNamespace, streamId.rawName)
+    override fun overwriteStage(streamId: StreamId, suffix: String) {
+        bigquery.delete(tableId(streamId, ""))
+        bigquery.query(
+            QueryJobConfiguration.of(
+                """ALTER TABLE `${streamId.rawNamespace}`.`${streamId.rawName}$suffix` RENAME TO `${streamId.rawName}`"""
+            ),
+        )
+    }
+
+    override fun transferFromTempStage(streamId: StreamId, suffix: String) {
+        // TODO figure out how to make this work
+        // something about incompatible partitioning spec (probably b/c we're copying from a temp
+        // table partitioned on generation ID into an old real raw table partitioned on
+        // extracted_at)
+        val tempRawTable = tableId(streamId, suffix)
+        //        val jobConf =
+        //            CopyJobConfiguration.newBuilder(tableId(streamId, ""), tempRawTable)
+        //                .setWriteDisposition(JobInfo.WriteDisposition.WRITE_APPEND)
+        //                .build()
+        //        val job = bigquery.create(JobInfo.of(jobConf))
+        //        BigQueryUtils.waitForJobFinish(job)
+
+        bigquery.query(
+            QueryJobConfiguration.of(
+                """
+                    INSERT INTO `${streamId.rawNamespace}`.`${streamId.rawName}`
+                    SELECT * FROM `${streamId.rawNamespace}`.`${streamId.rawName}$suffix`
+                """.trimIndent()
+            )
+        )
+        bigquery.delete(tempRawTable)
```

Comment:
wherever we are dropping table,  better to add safety check that suffix.isNotBlank so we don't accidentally delete the actual table ? because the contract of suffix is not-null but could be blank (even though the CDK code probably has safety check ?)

Reply:
any objection to adding an `if (suffix == "") throw IllegalArgumentException`? since that's presumably a CDK bug, i.e. we want tests to fail loudly and obviously

Reply:
not opposed to it in methods like these where suffix is required rather than used as both with and without suffix (like createTable).

Reply:
added checks to the overwrite/transferStage methods (and also swapped the overwriteFinalTable method to throw IllegalArgument)